test/tests.spec.ts

```diff
@@ -3256,296 +3086,256 @@ describe("passport-saml /", function () {
-              issuer: "http://localhost:20000/saml2/idp/metadata.php",
-              nameID: "stavros@workable.com",
-              nameIDFormat: "urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress",
-              sessionIndex: "_00bf7b2d5d9d3c970217eecefb1194bef3362a618e",
-            });
-            done();
-          } catch (err2) {
-            done(err2);
-          }
-        });
+describe("validateRedirect()", function () {
+  describe("idp slo", function () {
+    let samlObj: SAML;
+    beforeEach(function () {
+      samlObj = new SAML({
+        cert: fs.readFileSync(__dirname + "/static/acme_tools_com.cert", "ascii"),
+        idpIssuer: "http://localhost:20000/saml2/idp/metadata.php",
       });
+      this.request = Object.assign(
+        {},
+        JSON.parse(fs.readFileSync(__dirname + "/static/idp_slo_redirect.json", "utf8"))
+      );
+      this.clock = sinon.useFakeTimers(Date.parse("2018-04-11T14:08:00Z"));
     });
-    describe("sp slo", function () {
-      let samlObj: SAML;
-      beforeEach(function () {
-        samlObj = new SAML({
-          cert: fs.readFileSync(__dirname + "/static/acme_tools_com.cert", "ascii"),
-          idpIssuer: "http://localhost:20000/saml2/idp/metadata.php",
-          validateInResponseTo: true,
-        });
-        this.request = Object.assign(
-          {},
-          JSON.parse(fs.readFileSync(__dirname + "/static/sp_slo_redirect.json", "utf8"))
+    afterEach(function () {
+      this.clock.restore();
+    });
+    it("errors if bad xml", async function () {
+      const body = {
+        SAMLRequest: "asdf",
+      };
+      try {
+        await samlObj.validateRedirectAsync(body, null);
+        false.should.eql(true, "validateRedirectAsync should fail");
+      } catch (err) {
+        // success
+      }
+    });
+    it("errors if idpIssuer is set and issuer is wrong", async function () {
+      samlObj.options.idpIssuer = "foo";
+      try {
+        await samlObj.validateRedirectAsync(this.request, this.request.originalQuery);
+        false.should.eql(true, "validateRedirectAsync should fail");
+      } catch (err) {
+        err!.message!.should.eql(
+          "Unknown SAML issuer. Expected: foo Received: http://localhost:20000/saml2/idp/metadata.php"
         );
-        this.clock = sinon.useFakeTimers(Date.parse("2018-04-11T14:08:00Z"));
-      });
-      afterEach(async function () {
-        this.clock.restore();
-        await samlObj.cacheProvider.removeAsync("_79db1e7ad12ca1d63e5b");
-      });
-      it("errors if bad xml", function (done) {
-        const body = {
-          SAMLRequest: "asdf",
-        };
-        samlObj.validateRedirect(body, null, function (err) {
-          try {
-            should.exist(err);
-            done();
-          } catch (err2) {
-            done(err2);
-          }
-        });
+      }
+    });
+    it("errors if request has expired", async function () {
+      this.clock.restore();
```

Comment:
This shouldn't be needed if we are resetting the clock at the end of other tests.

Reply:
looks like this test needs real date, that's why it resets clock before